From 2b4e9b201abb859e8bf6717ce206d51efccf91e3 Mon Sep 17 00:00:00 2001
From: Aishwarya R <quic_aisr@quicinc.com>
Date: Tue, 28 Mar 2023 18:53:40 +0530
Subject: [PATCH] ath12k: Fixed reboot assert issue when MLO enabled

Facing firware assert issue as pci shutdown not
happening properly. ath12k_hif_power_down is responsible
for sending mhi_power_off state to firmware by this
it's power off one chip while the core does not stop
for other parallel running chips. calling core_deinit
during shutdown to fix this issue

Signed-off-by: Aishwarya R <quic_aisr@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/pci.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/pci.c b/drivers/net/wireless/ath/ath12k/pci.c
index 6d15da4..98b254b 100644
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -1434,7 +1434,9 @@ static void ath12k_pci_shutdown(struct pci_dev *pdev)
 {
 	struct ath12k_base *ab = pci_get_drvdata(pdev);
 
-	ath12k_pci_power_down(ab);
+	set_bit(ATH12K_FLAG_UNREGISTERING, &ab->ag->dev_flags);
+        cancel_work_sync(&ab->reset_work);
+        ath12k_core_deinit(ab);
 }
 
 static __maybe_unused int ath12k_pci_pm_suspend(struct device *dev)
-- 
2.17.1

